js/lex.go

```diff
@@ -161,6 +161,9 @@ func (l *Lexer) Next() (TokenType, []byte) {
 	// ErrorToken, WhitespaceToken and CommentToken are already returned
 	if tt == LineTerminatorToken || tt == PunctuatorToken && regexpStateByte[c] {
 		l.regexpState = true
+	} else if tt == IdentifierToken {
+		hash := ToHash(l.r.Lexeme())
+		l.regexpState = hash != 0 && hash != This
```

Comment:
I'm not sure yet which hashes could be added, I don't expect much though. However, if only a minority of the keywords could be followed by a regex I agree it is better to make a single if-statement (or switch, their performance is equivalent I believe).

So far I can see that the following keywords should match: `return`, `yield`, `throw`, `typeof`, `instanceof`, `in`, `default`, `case`? I'm not sure about all of them, but `if`, `do`, `for`, etc. would always first start with a `(` and thus would not need to be matched. `void` or `delete` or `extends` I don't think they are ever followed by a regex.